components/layout/query.rs

```diff
@@ -214,20 +209,62 @@ pub fn process_resolved_style_request(
                 let margins = box_fragment.margin;
                 let padding = box_fragment.padding;
                 let specific_layout_info = box_fragment.specific_layout_info.clone();
-                (content_rect, margins, padding, specific_layout_info)
+                let cumulative_containing_block_rect =
+                    box_fragment.cumulative_containing_block_rect;
+                (
+                    content_rect,
+                    margins,
+                    padding,
+                    specific_layout_info,
+                    cumulative_containing_block_rect,
+                )
             },
             Fragment::Positioning(positioning_fragment) => {
                 let content_rect = positioning_fragment.borrow().rect;
+                let cumulative_containing_block_rect = positioning_fragment
+                    .borrow()
+                    .cumulative_containing_block_rect;
                 (
                     content_rect,
                     SideOffsets2D::zero(),
                     SideOffsets2D::zero(),
                     None,
+                    cumulative_containing_block_rect,
                 )
             },
             _ => return computed_style(Some(fragment)),
         };
 
+        // For line height, the resolved value is the computed value if it
+        // is "normal" and the used value otherwise.
+        if longhand_id == LonghandId::LineHeight {
+            let font = style.get_font();
+            let font_size = font.font_size.computed_size();
+            return match font.line_height {
+                // There could be a fragment, but it's only interesting for `min-width` and `min-height`,
+                // so just pass None.
+                LineHeight::Normal => computed_style(None),
+                LineHeight::Number(value) => (font_size * value.0).to_css_string(),
+                LineHeight::Length(value) => value.0.to_css_string(),
+                LineHeight::MozBlockHeight => {
+                    let mut content_rect = content_rect;
+                    let writing_mode = style.writing_mode;
+                    let base_fragment = fragment.base().expect(
+                        "Base Fragment should exist for BoxFragment or PositioningFragment",
+                    );
+
+                    // For a inline level fragment it would be the containing block of it.
+                    // But, we shouldn't need to find the writing mode, since it will inherit.
+                    if style.is_inline_box(base_fragment.flags) {
```

Comment:
The comment says "inline level", the code checks for inline box. What about inline-level boxes which aren't inline boxes (e.g. inline-block), how should `-moz-block-height` behave?

Also the comment seems a bit weird. We don't support writing modes yet, but the writing mode could presumably be different, it's not guaranteed to inherit.

Reply:
I think `-moz-block-height` is only for internal usage and shouldn't be exposed to web authors, so we don't need to worry about atomic inlines. And we can simply add an assertion here that the box is just a regular inline box and not an atomic inline. For the same reason, the writing mode will not be different.

Btw we should have a test case to verify that `-moz-block-height` is not exposed.

Reply:
The problem is that Firefox only sets it to an internal pseudo-element. But this PR is setting it directly on the `<input>`, which is forced to be atomic when is has `display: inline` (#35908).

So it seems to me this needs a properly defined behavior for atomic inlines.

Reply:
Wait, I thought we have a UA shadow DOM under `<input>`, but it seems not?

What's the current box tree of an `<input>`?

Reply:
There's no shadow tree for `input` elements currently. They have some "magic" helper functions like `value_for_layout` and layout always renders them as text, ignoring any descendants (even `type=checkbox`, those are implemented in CSS :P).